src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/HACK_PasteOrdering.cs

```diff
@@ -0,0 +1,148 @@
+﻿using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Input.Commands.Ordering
+{
+    /// <summary>
+    /// This is a hack due to CPS not exposing easier ways to handle pasting items.
+    /// This does proper file ordering for pasting or dropping items into a folder or project.
+    /// </summary>
+    [Export(typeof(IPasteDataObjectProcessor))]
+    [Export(typeof(IPasteHandler))]
+    [AppliesTo(ProjectCapabilities.SortByDisplayOrder)]
+    [Order(OrderPrecedence)]
+    internal class HACK_PasteOrdering : IPasteHandler, IPasteDataObjectProcessor
+    {
+        public const int OrderPrecedence = 10000;
+
+        private readonly IPhysicalProjectTree _projectTree;
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+
+        private IProjectTree _dropTarget;
+
+        private IPasteHandler _pasteHandler;
+        private IPasteDataObjectProcessor _pasteProcessor;
+
+        [ImportingConstructor]
+        public HACK_PasteOrdering(IPhysicalProjectTree projectTree, IUnconfiguredProjectVsServices projectVsServices, UnconfiguredProject unconfiguredProject)
+        {
+            Requires.NotNull(projectTree, nameof(IPhysicalProjectTree));
+            Requires.NotNull(unconfiguredProject, nameof(UnconfiguredProject));
+
+            _projectTree = projectTree;
+            _projectVsServices = projectVsServices;
+
+            PasteHandlers = new OrderPrecedenceImportCollection<IPasteHandler>(projectCapabilityCheckProvider: unconfiguredProject);
+            PasteProcessors = new OrderPrecedenceImportCollection<IPasteDataObjectProcessor>(projectCapabilityCheckProvider: unconfiguredProject);
+        }
+
+        [ImportMany]
+        private OrderPrecedenceImportCollection<IPasteHandler> PasteHandlers { get; set; }
+
+        [ImportMany]
+        private OrderPrecedenceImportCollection<IPasteDataObjectProcessor> PasteProcessors { get; set; }
+
+        private IPasteHandler PasteHandler
+        {
+            get
+            {
+                if (_pasteHandler == null)
+                {
+                    // Grab the paste handler that has the highest order precedence that is below HACK_PasteOrdering's order precedence. 
+                    _pasteHandler = 
+                        PasteHandlers.Where(x => x.Metadata.OrderPrecedence < OrderPrecedence)
+                        .OrderByDescending(x => x.Metadata.OrderPrecedence).First().Value;
+                }
+
+                Assumes.NotNull(_pasteHandler);
```

Comment:
This is redundant, `.First()` throws if it is not found.

Reply:
Yes.